src/lib.rs

```diff
@@ -19,288 +23,241 @@ use syntax::ast_map;
+        //  - then add pat.id to self.map so we can track it going forward
+        // We also need to handle if l.source is a LocalFor
+
+        // self.cx.tcx.sess.span_note(d.span, &format!("decl: {:?}\n", d));
         if let DeclLocal(ref l) = d.node {
+            debug!("decllocal: {:?}\n", ty::pat_ty(self.cx.tcx, &l.pat));
             if l.source == LocalFor {
-                // We don't handle for bindings yet
-                // since they are tied to a different block
-                // Delegate to the usual walker
-                visit::walk_decl(self, d);
-                return;
+                unimplemented!();
             }
-            // First walk down the initializer
+
+            // Remove moved variables from map
+            // Maybe it's a reference? Use maybe_walk_expr
             if let Some(ref ex) = l.init {
                 self.visit_expr(&*ex);
             }
-            // grumble grumble Copy grumble
-            let block = self.current_block;
-            self.walk_pat_and_add(&*l.pat, block);
-        } else {
-            // Walk normally
-            visit::walk_decl(self, d);
+
+            // Add the ids in the pat
+            self.walk_pat_and_add(&*l.pat);
+            return;
+        }
+        visit::walk_decl(self, d);
+    }
+
+    fn visit_stmt(&mut self, s: &'v Stmt) {
+        if let StmtSemi(ref e, id) = s.node {
+            if is_protected(self.cx.tcx, e) {
+                self.cx.tcx.sess.span_err(s.span, "Return type is protected but unused");
+            }
         }
+        visit::walk_stmt(self, s);
     }
 
-    fn visit_expr(&mut self, ex: &'v Expr) {
-        match ex.node {
-            // Might need to be ExprPath(None, _)
-            // to work on current nightly
+    fn visit_expr(&mut self, e: &'v Expr) {
+        // Visit and remove all consumed values
+        // Which Exprs do we need to handle?
+        // At least ExprCall and ExprMethodCall
+        debug!("visit_expr: {:?}\n", e);
+        match e.node {
+            ExprAssign(ref e1, ref e2) => {
+                // Remove all protected vars in rhs
+                visit::walk_expr(self, &e2);
 
-            ExprPath(_, _) => {
-                let def = self.cx.tcx.def_map.borrow().get(&ex.id).map(|&v| v);
-                if let Some(PathResolution { base_def: DefLocal(id),  .. }) = def {
-                    if is_protected(self.cx.tcx, ex) {
-                        let decl = self.map[id];
-                        // TODO use proper lint erroring
-                        self.cx.tcx.sess.span_warn(ex.span, "found usage of variable of protected type");
-                        self.cx.tcx.sess.span_note(decl.1, "declaration here");
-                    }
+                // TODO: Add all protected vars in lhs
+                if let ExprPath(_, _) = e1.node {
+                    self.map.insert(expr_to_localid(self.cx.tcx, e1).unwrap(), e.span);
```

Comment:
My test-case was something like `let mut x = Foo; x = x`, variants of which we use quite a bit in our session code.